### ISSUE TITLE
Question: [js-flipper] supported device types, localhost connection, unique device id

### ISSUE THREAD

**Issue by usrbowe** [facebook/flipper#3319]
I've been testing the new `js-flipper`, and was planning to have a way to connect RN apps, without the native Flipper SDK (this might be also the way Expo could adopt Flipper).

There are few problems so far:

1. the `os` package is no longer polyfilled by [webpack](https://github.com/webpack/changelog-v5#automatic-nodejs-polyfills-removed) (so also latest create react app will not be able to use the `js-flipper`).

2.  supported devices`Browser` or `Node.JS` is quite limiting factor. I think we could have a way to pass on device type when we run `flipperClient.start` to tell Flipper how to display the device and list supported plugins.
https://github.com/facebook/flipper/blob/11c3a20190a972a0d7a804124b6d72bc8d3afb0f/js/js-flipper/src/util.ts#L49-L60

~~3. Valid websocket connection prefixes will only accept `'chrome-extension://', 'localhost:', 'http://localhost:', 'app://'`. Which makes it impossible to connect from other device. Actually would like to know more details on why this limitation exist, maybe could have some opt out or other way to accept connections from non localhost.~~
For Node/RN can provide custom `origin: localhost` header.

https://github.com/facebook/flipper/blob/11c3a20190a972a0d7a804124b6d72bc8d3afb0f/desktop/flipper-server-core/src/comms/BrowserServerWebSocket.tsx#L150-L152

4. When running `react-flipper-example`, every time the connection is established there will be new device added and previous connection marked s `Offline`. I think we might need have some way to assign same browser/apps to existing connection. 
<img width="256" alt="image" src="https://user-images.githubusercontent.com/5339061/150054660-55131d4b-3af2-4149-b628-83406d569887.png">

**Comment by usrbowe** [comment]
3. **IDEA**: Actually for the `non-localhost` connections, would it be viable to provide QR code mechanism? To first get the Flipper host IP address and probably we could also exchange some passphrase to establish secure connection.

**Comment by usrbowe** [comment]
@aigoncharov thanks for reply, somehow missed this. I forgot to point that I'm testing this connection for simpler React Native connection (without the native Flipper SDK).

#### 1. package `os` not available
`TODO`: will provide more details, I did tested on fresh CreateReactApp.
On React Native, it seems metro will still execute the `require` by default (but there seems to be option to lazily invoke require).


### 2. Support devices
Actually, the `deviceType` would be only needed for device plugins (for other plugins, it would be invoked by addPlugin method).

####  3. WS `localhost` validation
I tested now again on React Native and realised it also have an option to specify `origin` headers (similar as `ws` package on NodeJS):
```js
// reference for other connecting React Native
// let metroServerIP = set IP address of metro server (by default: localhost, for physical devices it will be the metro server host IP)
const flipperPort = 8333;
flipperClient.start('React Native App', {
  urlBase: `${metroServerIP}:${flipperPort}`,
  websocketFactory: url =>
    new WebSocket(url, undefined, {
      headers: {
        origin: `localhost:${flipperPort}`,
      },
    }),
});
```

So feels actionable items could be only:

1. Add `getDeviceId: () => Promise<string>` to restore previous session
2. Easier connection for Flipper (also allow discovery of devices/apps so don't have to rely on adb/idb)

**Comment by usrbowe** [comment]
> The question here is which environment are we discussing? If it is a browser, what would be the case when a non-localhost client needs to connect to Flipper? If it is not a browser, then couldn't it simply pass "Origin" header set to "localhost" to get past the validator?

I think we could probably provide QR code on Flipper app, which devices could scan to at least get the IP address of host. Also this could cover the secure connection exchange.

**Comment by aigoncharov** [comment]
Polyfilling issue for the `os` should be resolved by 245d263c1e7beb34a42a65fa80f1045d47cbee24

**Comment by aigoncharov** [comment]
`getDeviceId` should be resolved by 055b14c6dd7110ac8d927ea07c32aeb9ba8bee3c
